danceradvent/public/articles/2020/23-asynchronous-promises.pod

```diff
@@ -0,0 +1,526 @@
+interface by default.
+
+    package CovidStats::IOAsync;
+
+    use Dancer2;
+    use experimental qw< postderef signatures >;
+
+    use DateTime;
+    use IO::Async;
+    use IO::Async::Loop;
+    use Future::Utils qw< fmap >;
+    use Net::Async::HTTP;
+    use Statistics::Descriptive::Full;
+    use URI::Escape qw< uri_escape >;
+
+    use constant {
+        'MAX_COUNTRIES' => 5,
+        'MAX_DAYS'      => 14,
+        'SUMMARY_URL'   => 'https://covid-api.mmediagroup.fr/v1/cases',
+        'COUNTRY_URL'   => 'https://covid-api.mmediagroup.fr/v1/history?country=%s&status=Confirmed',
+        'LTRIM'         => '0.25',
+    };
+
+    get '/' => sub {
+        return delayed {
+            flush();
+
+            my $loop = IO::Async::Loop->new();
+            my $http = Net::Async::HTTP->new();
+            $loop->add($http);
+
+            my $main_req = $http->GET( SUMMARY_URL() )->then( sub ($res) {
+                my $data      = from_json( $res->content );
+                my @countries = ( sort {
+                    $data->{$b}{'All'}{'confirmed'} <=> $data->{$a}{'All'}{'confirmed'}
+                } grep $_ ne 'Global', keys $data->%* )[ 0 .. MAX_COUNTRIES() ];
+
+                if ( !@countries ) {
+                    content("Sorry, failed to received countries\n");
+                    done();
+                    return;
+                }
+
+                return \@countries;
+            })->then( sub ($countries) {
+                my $yesterday = DateTime->now->subtract( 'days' => 1 );
+
+                fmap( sub ($country) {
+                    my $data_url = sprintf COUNTRY_URL(), uri_escape($country);
+
+                    return $http->GET($data_url)->then( sub ($res) {
+                        my $country_data = from_json( $res->content );
+
+                        my @period;
+                        my $day = $yesterday;
+                        foreach ( 0 .. MAX_DAYS() ) {
+                            push @period, $day->ymd();
+                            $day = $day->subtract( 'days' => 1 );
+                        }
+
+                        my @period_data = $country_data->{'All'}{'dates'}->@{@period};
+
+                        my $stat = Statistics::Descriptive::Full->new();
+                        $stat->add_data($_)
+                            for @period_data;
+
+                        return [ $country => $stat->trimmed_mean( LTRIM() ) ];
+                    })->catch( sub ($error) {
+                        content("Sorry, failed to fetch data for $country: $error");
+                    });
+                }, 'foreach'    => $countries,
+                   'concurrent' => scalar $countries->@*,
+                )->then( sub (@results) {
+                    content( "By country (period of " . MAX_DAYS() . " days):\n" );
+                    content( "- $_->[0]: $_->[1]\n" )
+                        for sort { $b->[1] <=> $a->[1] } @results;
```

Comment:
You didn't want to

```perl
use List::UtilsBy qw< nsort_by >;
...

    for nsort_by { $_->[1] } @results;
```
?

Reply:
Someone not familiar with `List::UtilsBy` might not know what this is, requiring another trip to the documentation to read and say "D'oh, it's just a simple numerical sort." I'd rather keep it to basics and avoid extra modules. (Says the man who took at least one entire hour to figure out a better statistical function than mean average and found the large statistics module just to provided a different metric and had to then include an explanation on it.)